### PR TITLE
In memory one off

### PR DESCRIPTION
**Pull request by wasade**
This pull request provides a new API method to allow computation on existing `biom` and `BPTree` structures, and additionally exposes a means to construct a `biom` from in memory objects. Note that `BPTree` already has support instantiation from in memory constructs.

### CODE REVIEW

File: src/api.cpp
Comment (wasade) [inline src/api.cpp:104]: I could not get the code to compile using `biom` rather than `biom_interface`, but I confess I'm not fully sure why

File: src/api.hpp
Comment (wasade) [inline src/api.hpp:176]: same here, I couldn't get `biom` to work but `biom_interface` did. However, I think this is why the test code is failing as this triggers the `include <vector>` that the C-API test is not happy about

File: src/biom.hpp
Comment (wasade) [inline src/biom.hpp:89]: we should test the number of entries in `indptr` to make sure its shape makes sense to the shape of `id_`

File: src/api.hpp
Comment (wasade) [inline src/api.hpp:2]: These bring in C++ headers, which break the C-API test. At the moment, I'm unsure how to correctly address this. I don't believe we can just wrap the `include`s in `#if def __cplusplus` in `api.hpp` as that would leave `biom` and `BPTree` undefined, and we cannot do it within `biom.hpp` and `tree.hpp` as that would leave components of their signatures undefined.

Comment (wasade) [reply src/api.hpp]: ...the present solution is to limit use of `one_off_inmem` to c++...

File: src/biom.cpp
Comment (sfiligoi) [inline src/biom.cpp:117]: I would suggest  explicitly initialize has_hdf5_backing=false, since several functions look for it.

File: src/api.hpp
Comment (sfiligoi) [inline src/api.hpp:216]: Do we really need to define it here?
Or would it be better as an internal function, only?

Comment (wasade) [reply src/api.hpp:216]: If we move it internal (perhaps to su.cpp?), then we need to migrate the definition of ComputeStatus to avoid a circular dependency. Any concern if we move ComputeStatus to a new header?

Comment (wasade) [reply src/api.hpp:216]: What I propose is we leave it here for now, on the rational that other non-exposed methods exist in api.cpp/hpp, get things working, and consider a refactor later. Does that sound reasonable? Part of the hesitation to solve immediately is it requires centralizing the currently used macros, or refactoring them into functions. Or rather, it doesn't seem like a straight forward reorganization. If you agree, I can look at it on a follow up PR. Does that sound okay?

Comment (wasade) [reply src/api.hpp:216]: nm, actually i think it is important to do this as its signature contains non-C types. I'll sort it out on this PR

File: src/api.hpp
Comment (sfiligoi) [inline src/api.hpp]: Do we still need these two?

File: src/Makefile
Comment (sfiligoi) [inline src/Makefile:138]: We need status_enum.hpp, since api.hpp imports it.

Comment (wasade) [reply src/Makefile:138]: +1

File: src/Makefile
Comment (sfiligoi) [inline src/Makefile:140]: Who needs biom.hpp, tree.hpp and biom_interface.hpp, if they are were not used in api.hpp?

Comment (wasade) [reply src/Makefile:140]: Good point, no longer needed

File: src/api.hpp
Comment (wasade) [inline src/api.hpp:207]: @sfiligoi does anything unusual stand out here? It looks like the compiler is not honoring the `extern "C"`:

```bash
$ nm -D libssu.so | grep _inmem
00000000000ca9f0 T _Z13one_off_inmemPK12support_biomPK14support_bptreePKcbdbjPP3mat
00000000000c8a20 T _Z17one_off_inmem_cppRN2su4biomERNS_6BPTreeEPKcbdbjPP3mat
```

As far as I know, both `support_biom_t` and `support_bptree_t` are entirely native C types. 

It's not that `extern` isn't working as some functions are extern'd unmangled:

```bash
$ nm -D libssu.so | grep "one_off"
00000000000ca180 T faith_pd_one_off
00000000000cab20 T one_off
00000000000cc270 T one_off_matrix
00000000000cd310 T one_off_matrix_fp32
00000000000ca9f0 T _Z13one_off_inmemPK12support_biomPK14support_bptreePKcbdbjPP3mat
00000000000cb1e0 W _Z16one_off_matrix_TId13mat_full_fp64E14compute_statusPKcS3_S3_bdbjS3_PPT0_
00000000000cc280 W _Z16one_off_matrix_TIf13mat_full_fp32E14compute_statusPKcS3_S3_bdbjS3_PPT0_
00000000000c8a20 T _Z17one_off_inmem_cppRN2su4biomERNS_6BPTreeEPKcbdbjPP3mat
```

I'm going to continue to tinker, but wanted to ping incase I was missing something obvious

Comment (wasade) [reply src/api.hpp:207]: Found the issue

**Review by sfiligoi — APPROVED** [review]
Looks good to me

File: src/biom.cpp
Comment (sfiligoi) [inline src/biom.cpp:137]: Performance note:
If n_objs is large, we may want to create the various vector in parallel.

File: src/biom.cpp
Comment (sfiligoi) [inline src/biom.cpp:135]: Performance note:
Instead of doing reserse + for{push_back},
You probably should do 
resize+memcopy
Likely way faster.

Comment (wasade) [reply src/biom.cpp:140]: I'm unsure how to resolve this if using `std::string` as the individual `char*`s would not be instantiated `std::string`. For the moment, I've wrapped this in a parallel block. I think the right thing to do here would be to remove use of `std::string`, which I'll examine when looking at removing copies.

File: src/api.cpp
Comment (sfiligoi) [inline src/api.cpp:377]: Since we are changing the Python interface anyhow, 
we may as well change the return value.

Getting the data in and out of the condense for is extremely expensive;
we should use the flat matrix form instead (that's that the python representation is anyhow).

I had not done it in the past, because I was under the assumption that the cmdline interface was the most performance critical one, but since we/you are now optimizing the python interface performance, this could buy you more than all of the other optimizations combined.

File: src/api.cpp
Comment (sfiligoi) [inline src/api.cpp:440]: You know the size of the resulting matrix at this point, so you would be better off to pass the pre-allocated buffer and fill it in, instead of letting one_off create one that you then need to make a copy of.
(even better if it was in plain-matrix, and not condensed format)

File: src/biom.cpp
Comment (wasade) [inline src/biom.cpp:129]: ...we probably could still use const but i think it would require updating the object definition to change the member variables themselves to be const

File: src/biom.cpp
Comment (wasade) [inline src/biom.cpp:438]: Please check this close, this is the first time I've done a reduction. While the tests pass, which improves confidence, I think a close examination of the use of parallel here is warranted.

File: src/biom.hpp
Comment (wasade) [reply src/biom.hpp:97]: ...we can't really do this, we *assume* `indptr` is of length n_obs + 1. This sanity check is the resposibility of the caller

Comment (wasade) [reply src/biom.hpp:97]: ...but also this method isn't needed anymore

File: src/biom.cpp
Comment (sfiligoi) [reply src/biom.cpp:439]: Are you sure this is even legal?
reduction needs a local copy of whatever needs to be reduced.

This code seems way to convoluted to be parallelized in a naive way.
I think it needs a complete refactoring, if we really want to parallelize it.
(is it really that expensive.?)

Comment (wasade) [reply src/biom.cpp:439]: I thought it was a gamble.... :) The tests pass, but that doesn't mean it is a safe approach.

It is plausibly more expensive than the other operations on load (other than malloc') as its a full table scan. To be fair, I have not profiled it. If you recommend backing this out, and we defer to later, that's totally fine

Comment (wasade) [reply src/biom.cpp]: Removed

File: src/biom.cpp
Comment (sfiligoi) [inline src/biom.cpp:312]: Why not use directly calloc?
It is shorter and probably faster.

File: src/test_su.cpp
Comment (sfiligoi) [inline src/test_su.cpp:156]: This is cosmetic, but would be nice to remove.

File: src/test_su.cpp
Comment (sfiligoi) [inline src/test_su.cpp:466]: Cosmetic, but useless diff.

**Comment by sfiligoi** [comment]
Looks OK, but see my 3 minor comments above.

File: src/biom.cpp
Comment (wasade) [reply src/biom.cpp:312]: Good point, left over from when attempting to parallelize this

**Comment by wasade** [comment]
Thanks! Just addressed I believe

Once green, I'll merge and see if/whats needed for bioconda

**Review by sfiligoi — APPROVED** [review]
Looks good now

**Comment by wasade** [comment]
Should be internal — thanks

**Pull request merged**
